rogue/src/blessings.rs

```diff
@@ -0,0 +1,94 @@
+
+use specs::prelude::*;
+use super::{
+    BlessingOrbBag, OfferedBlessing, BlessingSelectionTile, Position, Point,
+    WantsToDissipate, InSpellBook, Castable, entity_spawners
+};
+
+pub fn create_offered_blessings(ecs: &mut World) {
+    let mut offers: Vec<Entity> = Vec::new();
+    let offer = entity_spawners::spells::magic_missile(ecs, 0, 0, 5, 2);
+    match offer {
+        Some(offer) => {
+            let mut offered_blessings = ecs.write_storage::<OfferedBlessing>();
+            offered_blessings.insert(offer, OfferedBlessing {})
+                .expect("Could not insert offered blessing component.");
+            offers.push(offer);
+        }
+        None => {}
+    };
+    let offer = entity_spawners::spells::fireball(ecs, 0, 0, 5, 2);
+    match offer {
+        Some(offer) => {
+            let mut offered_blessings = ecs.write_storage::<OfferedBlessing>();
+            offered_blessings.insert(offer, OfferedBlessing {})
+                .expect("Could not insert offered blessing component.");
+            offers.push(offer);
+        }
+        None => {}
+    };
+    let offer = entity_spawners::spells::icespike(ecs, 0, 0, 5, 2);
+    match offer {
+        Some(offer) => {
+            let mut offered_blessings = ecs.write_storage::<OfferedBlessing>();
+            offered_blessings.insert(offer, OfferedBlessing {})
+                .expect("Could not insert offered blessing component.");
+            offers.push(offer);
+        }
+        None => {}
+    };
+    ecs.maintain();
+}
+
+// Add a blessing to the player's spellbook.
+pub fn receive_blessing(ecs: &mut World, blessing: Entity) {
```

Comment:
It's hard assumed here that all blessings are castable. To generalize, we will need to classify blessings into types and write conditional logic based on the type.

Reply:
Would that be using something similar to `(&entities, &castables).join()` for each class of thing? I haven't dug too deeply into how that works under the covers yet, but I assume based on the mention of using a bitset that it's reasonably efficient.

Reply:
I haven't really thought about how to implement it, I'm honestly not quite convinced this sort of progression system is a great idea. I mostly stole the idea from Hades.

I think the conditional logic would branch on like, how the boon is "applied" to the player. Like, here, we have this `InSpellbook` component, and you attach it to the spell with a pointer to the player. If the boon was a buff, or some other static effect, we would need something comparable.

We should think about some way to model buffs, nerfs. I remember being unsatisfied with it.